R/time_converters.R

```diff
@@ -0,0 +1,127 @@
+#' convert an archive from daily data to weekly data, summing where appropriate
+#' @details
+#' this function is slow, so make sure you are calling it correctly, and
+#'   consider testing it on a small portion of your archive first
+#' @param day_of_week integer, day of the week, starting from Sunday, select the
+#'   date to represent the week in the time_value column, based on it's
+#'   corresponding day of the week. The default value represents the week using
+#'   Wednesday.
+#' @param day_of_week_end integer, day of the week starting on Sunday.
+#'   Represents the last day, so the week consists of data summed to this day.
+#'   The default value `6` means that the week is summed from Sunday through
+#'   Saturday.
+daily_to_weekly <- function(epi_arch,
+                            agg_columns,
+                            agg_method = c("sum", "mean"),
+                            day_of_week = 4L,
+                            day_of_week_end = 6L) {
+  agg_method <- arg_match(agg_method)
+  if (agg_method == "total") {
+    agg_fun <- epi_slide_sum
+  } else if (agg_method == "mean") {
+    agg_fun <- epi_slide_mean
+  }
+  keys <- grep("time_value", key_colnames(epi_arch), invert = TRUE, value = TRUE)
+  too_many_tibbles <- epix_slide(
+    epi_arch,
+    before = 99999999L,
+    ref_time_values = ref_time_values,
+    function(x, group, ref_time) {
+      x %>%
+        group_by(across(all_of(keys))) %>%
+        agg_fun(agg_columns, before = 6L) %>%
+        select(-all_of(agg_columns)) %>%
+        rename_with(~ gsub("slide_value_", "", .x)) %>%
+        # only keep 1/week
+        filter(wday(time_value) == day_of_week_end) %>%
+        # switch time_value to the designated day of the week
+        mutate(time_value = time_value - 7L + day_of_week) %>%
```

Comment:
issue?: this might need to be
```suggestion
        mutate(time_value = time_value - 7L + (day_of_week - day_of_week_end) %% 7L) %>%
```
Probably easiest to construct a test case with `day_of_week_end` not Saturday & see if anything gives the right result.

Reply:
Did a little test, seems to back up current suggestion above.
``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(tidyr)
crossing(end_date = as.Date("2020-01-01") + 0:6, day_of_week = 1:7) %>%
  mutate(day_of_week_end = lubridate::wday(end_date, week_start = 7L)) %>%
  mutate(
    method1 = end_date - 7L + day_of_week,
    method2 = end_date - 7L + (day_of_week - day_of_week_end) %% 7L,
    method3 = end_date - (day_of_week_end - day_of_week) %% 7L
  ) %>%
  pivot_longer(method1:method3, names_to = "method", values_to = "date") %>%
  mutate(day_of_week_actual = lubridate::wday(date, week_start = 7L),
         right_day_of_week = day_of_week == day_of_week_actual,
         right_week = as.integer(end_date - date) %in% 0:6) %>%
  summarize(.by = "method", issues = sum(!right_day_of_week | !right_week))
#> # A tibble: 3 × 2
#>   method  issues
#>   <chr>    <int>
#> 1 method1     42
#> 2 method2      7
#> 3 method3      0
```

<sup>Created on 2024-09-28 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
yeah this definite seems right, thanks for the catch